### ISSUE TITLE
✨ Cobalt 2.0

### ISSUE THREAD

**Issue by drwpow** [drwpow/cobalt-ui#201]
## Planning for Cobalt 2.0 is underway!

After many wonderful people have put Cobalt through the ringer, and have liked what it can do, but still need more from it, I’ve started planning for an upcoming breaking change release. As of right now, the high points include:

### Major changes

- **Ability to chain plugins.** Right now plugins are completely isolated from one another (except for CSS + Sass, which have rough edges). This has been workable with official plugins, but imposes restrictions on how custom plugins are made.

  I’m currently working on a draft for the Plugin 2.0 syntax, which is basically “the same, but more.” It will give plugins (many) more hooks than just the one `build()` hook so they can build off one another. And while it’s still a little early for feedback, I’ll open up RFCs in the coming weeks when I’ve proven the changes will simplify current plugins _and_ allow for more complex chaining and better sharing of logic between plugins. The API will still be primarily Rollup/Vite plugin-inspired, but will introduce new concepts where it makes sense (since tokens are fundamentally different than web bundlers)
- **Support for DTCG 1.0.** There has been no _official_ announcement of this yet, and no timeline, but rumors have it that DTCG 1.0 will be announced later this year. I’d like to take the opportunity for a breaking plugin API change to also support DTCG 1.0 as well. Also, if DTCG announces some sort of release cycle, I’d also like to start planning Cobalt releases around that as well.

### Minor changes

- **Parsed Tokens object instead of an array.** The initial version parsed `tokens.json` and returned a flat array of tokens, but this will be changed to be a shallow object (with `.` in keys).

  The original assumption was “since plugins will always be iterating over all tokens anyway, let’s make it simpler,” but that has proven false over time. There are many times when plugins need to grab individual tokens, and `tokens.find(…)` is not only verbose but slow (comparatively). Objects are closer to the original `tokens.json`, which makes things simpler to work with. The parsed tokens object will still be shallow, because it is still more time-consuming to do recursive AST walking than it is to run through a loop once.

### Staying the same (no changes)

- **Performance is (still) a priority.** I am a huge fan of Rust and wasm, but the nature of working with design tokens is still well within the realm of what Node.js does well. And adopting any additional layer of the stack could potentially make contributing, debugging, and plugin dev harder. I’m always conscious of how fast (or slow) Cobalt is, but as of now it’s very fast and lightweight, and 2.0 won’t change that.
- **Ease of creating plugins is (still) a priority.** Above all, it should be great DX to use Cobalt with existing plugins, and shouldn’t be much harder to write your own. Rollup/Vite continue to be sources of inspiration of timeless APIs that are the perfect blend of simplicity where needed, and complexity where required.
- **No new token types** (for now). “Boo!”—_I know_. This is actually good news, in a sense. If  DTCG 1.0 releases this year, it will be a big deal! Figma, Adobe, PenPot, and more have eyes on the DTCG spec, and plan to implement it in their products as soon as possible. Cobalt exists first and foremost to implement that spec as-designed, which maximizes the number of ways you can use Cobalt and its plugins. It will be better for everyone to not have [yet another competing standard](https://xkcd.com/927/).
- **(Still) independent of Style Dictionary**. Since Cobalt’s release, Style Dictionary has announced an [exciting new v4 with support for DTCG](https://github.com/amzn/style-dictionary/issues) (still unreleased as of this issue creation). This project doesn’t exist to compete with SD at all; on the contrary, if any design tooling succeeds, we all succeed! Cobalt only exists to give options (and allow preference). As a developer, I still prefer plugin-centric APIs for design tooling over than a looser, config-heavy “[Gulp](https://en.wikipedia.org/wiki/Gulp.js)-like” approach that SD takes. There is no right or wrong here; only freedom of expression! And I still see the value (and am having fun) trying out a philosophically-different approach from what SD is doing, while wishing that project all the best, and encouraging everyone that doesn’t fit with Cobalt to use SD for their needs.

_Note: this is a living document; expect edits (and even some decision reversals) as plans develop and mature._

## Release date: TBD, with RFCs coming Mar/Apr
 
To recap, this is mostly a placeholder announcement to check back in at a future date for RFCs and testable beta versions of Cobalt 2.0. While a stable version won’t release before DTCG 1.0 is closer, look out for both RFCs and testable beta versions very soon (in the coming weeks).

So all that to say, **is there anything you’d like to see in Cobalt 2.0?** At this early stage, any and all ideas are welcome (keeping in mind a Plugin 2.0 API will get an RFC soon)! Is there more Cobalt could do out-of-the-box? Any work that could be moved out of your plugin? Any additional integrations you’d like to see? Leave a comment!

Thanks for reading 🙏

**Comment by drwpow** [comment]
😄 Fitting that this is [201](https://http.cat/status/201)

**Comment by torstendaeges** [comment]
Sounds great, really looking forward to this! :)

**Comment by drwpow** [comment]
Had a bit of a breakthrough this week in what’s probably the most exciting development since this library started: **token linting**. As a PoC I’ve shipped an [a11y plugin](https://cobalt-ui.pages.dev/integrations/a11y) that can perform contrast checks on your tokens automatically (WCAG 2 and APCA). And under-the-hood is a whole new linting setup that lets Cobalt plugins register linters that users can configure, similar to ESLint.

This is huge because Figma plugins exist, but are manual. And designers shouldn’t perform QA (nor does the “source of truth” usually exist in Figma anyway). a11y tests also exist of course, but are often very slow and expensive because without a token pipeline you have to run full E2E tests with browsers (and still miss some things). But with **token linting,** the tests are fast, cheap, accurate, and surface in the part of the chain where they can be fixed: design ops.

I’m excited for all the possibilities this opens, especially with the new improvements coming in 2.0. And would love to hear ideas for other linting checks we could add to test the concept further!

**Comment by dzonatan** [comment]
> And would love to hear ideas for other linting checks we could add to test the concept further!

not related to a11y but it would be nice to check if there are no duplicate token names, e.g.: 
```json
      "blue": {
        "0": { "$value": "#ddf4ff" },
        "1": { "$value": "#b6e3ff" },
        "2": { "$value": "#80ccff" },
        "3": { "$value": "#54aeff" },
        "4": { "$value": "#218bff" }, // <--
        "4": { "$value": "#0969da" }, // <--
        "5": { "$value": "#0969da" },
        "6": { "$value": "#0550ae" },
        "7": { "$value": "#033d8b" },
        "8": { "$value": "#0a3069" },
        "9": { "$value": "#002155" }
      }
```
now last value wins. maybe this could be even checked in compile time.
IDEs usually flag these duplicate properties but this could happen by other means like wrong git conflict resolution.

also, just a thought, maybe checking duplicate token values within the same group would also be helpful:
```json
      "blue": {
        "0": { "$value": "#ddf4ff" },
        "1": { "$value": "#b6e3ff" },
        "2": { "$value": "#80ccff" },
        "3": { "$value": "#54aeff" },
        "4": { "$value": "#218bff" }, // <--
        "5": { "$value": "#218bff" }, // <--
        "6": { "$value": "#0550ae" },
        "7": { "$value": "#033d8b" },
        "8": { "$value": "#0a3069" },
        "9": { "$value": "#002155" }
      }
```

**Comment by drwpow** [comment]
Yes @dzonatan that’s exactly what I’m talking about! Duplicate values could be done today. Duplicate keys (IDs) would have to wait for 2.0 as that requires AST awareness (parser rebuild). But both are good examples of “core” lint rules that I think should just work out-of-the-box (maybe the lint rules are opt-in, but no plugins to install)

**Comment by drwpow** [comment]
@dzonatan Thanks to your suggestion, I just shipped the `duplicate-values` lint rule (along with some others) that lets you catch duplicate values anywhere in your tokens.json file. Please [give it a try](https://cobalt-ui.pages.dev/guides/linting) and let me know if you encounter any errors 🙏

duplicate _keys_ turned out to be pretty tricky! Because the JSON spec does allow for duplicate keys (it just takes the last one silently), fixing that requires a implementing a custom JSON parser so that may take a while 😅

**Comment by drwpow** [comment]
# 💠 Announcing … Terrazzo!

Figured I’d let the cat out of the bag with future plans for this library in the 2.0 release. I won’t give all the details here, but just some high points:

## Cobalt will be renamed to Terrazzo!

Cobalt was always a placeholder name. I was always more focused on the actual tooling for DTCG, and never felt truly attached to the name. “Cobalt” is a common name that competes with many things, both in programming/design and beyond. I never felt it captured the nature of tokens: these little beautiful, colorful things that come together to form this wonderfully-intricate and unique pattern…

…kinda like terrazzo!

The name also highlights an expansion in the maintainers. For a while now, Cobalt has been a solo side-project. But this year I’ve roped in a good friend and incredibly-talented designer, @ntassone, to make this 100× better than it ever would be with just myself. Together, we’ve  realized a lot of pieces were missing in the story of design tokens:

- **Visual Tokens Editor**: edit tokens with an OSS tool (kinda like Storybook but for tokens)
- **Better design tools**: I’ve been inspired by a lot of the work that Evil Martians have done such as [oklch.com](https://oklch.com), and wanted to [make my own](https://terrazzo-storybook.pages.dev/?path=/story/components-form-colorpicker--overview) 
- **Token linting**: I realized this could exist today, so [I built it this month](https://cobalt-ui.pages.dev/guides/linting), but have more ideas down this way
- **A bigger ecosystem**: I’ve been working on Cobalt solo for a few years now as a passion project, and I want to do things that incorporate more people (like @ntassone)!

So all that considered, I wanted a bigger, better Terrazzo namespace to contain all these ideas and experiments we’ve been working on for a while behind-the-scenes. And also open the doors to more people getting involved.

## Is this a company?

No. It’s still an 100% open-source project @ntassone and I are working on in our free time. It’d be neat to find a way to get more funding for this somehow! But only time will tell.

## So is Cobalt going away?

Nope! Cobalt v2 == Terrazzo CLI v1. 100% of Cobalt’s direction, vision, and philosophy won’t change. Terrazzo will just be a project that’s a suite of token tools in addition to _just_ a CLI.

## Why announce it here? … In this thread?

The Terrazzo nameswap was something I had been planning before even announcing 2.0. But @ntassone and I had been mulling over what the actual scope was until pretty recently, and how Terrazzo related to Cobalt. After deciding on a scope, and deciding Terrazzo was the same as Cobalt, I knew the actual swapover would be a little clumsy, and there wouldn’t be some seamless overnight launch. Over the next few months, you’ll start seeing the name “Terrazzo” here and there in the project. And one (undisclosed) day in the future, this repo will suddenly move to the Terrazzo GitHub org. And I wanted people to know it’s not changing hands (just adding some!).

I actually don’t even like talking about things before I do them. Doing is more fun. But the people reading this thread would start to notice the changes before anyone else, so this seems like the most effective way to announce it 🙂.

And as always, always open to any questions people have! 🙇

**Comment by drwpow** [comment]
Lil sneak peak: the beta version is a little further away than planned (hoping to get a version published in a couple weeks, but we’ll see), but here’s a sneak peak: full AST support for tokens.json! 😎 

![CleanShot 2024-04-09 at 09 53 07@2x](https://github.com/drwpow/cobalt-ui/assets/1369770/4168a1f8-1f7e-46b3-a5ce-47d871864e25)

It’s not _just_ showing line errors (though it does enable that). It also exposes the JSON AST to plugins as well so you’ll have access to the **original `tokens.json` file as-authored** (if you need it).

This will also enable line-specific lint messages as well! Excited for this update!

**Comment by drwpow** [comment]
Lil update: the 2.0 version is very, very alpha, but those interested can take a look at the new 2.0 docs here regarding the plugin changes: https://terrazzo-docs-beta.pages.dev/docs/cli/api/plugin-development/ (don’t @-me the docs are still largely unstyled, and this is just an early preview). Design credit for branding goes to @ntassone; I couldn’t have done this without him!

You can also kick the tires on `@terrazzo/cli` and `@terrazzo/plugin-css` which are available on npm. Work the same way, but with a `terrazzo.config.js` file instead of `tokens.config.js`.

With the core rewrite mostly done, hoping to have the rest of the plugins upgraded to 2.0 in the next month, finish styling / writing docs, and hopefully we’ll have some release candidates ready in the coming weeks.

And as always, feedback is always welcome! (particularly about the [plugin API enhancements](https://terrazzo-docs-beta.pages.dev/docs/cli/api/plugin-development/))

**Comment by drwpow** [comment]
#### Nov 2024 update

I’ve been working closely with the [DTCG](https://github.com/design-tokens/community-group) to finalize the last few things for the stable version of the spec. I still want to wait until some of the bigger changes are settled (there’s a **BIG** update coming for modes—get excited! 👀). Until then, Terrazzo supports the last stable version of DTCG, as well as the in-progress draft spec as changes happen.

While we’re all waiting for that, the Terrazzo CLI + plugins are fully usable today, and are already used in some production systems (there are a few minor breaking changes, but for most people using standard plugins it should be [an easy upgrade](https://terrazzo.app/docs/cli/migrating/)). 

And we’re also working on the Tokens Lab, to be a playground for playing with token files and start from existing design systems. That idea has always been part of the thinking for this project; other things just got in the way. But it’s actively being developed now with some help from friends.

If you want to track how things are progressing, there’s [a new GitHub Project](https://github.com/orgs/terrazzoapp/projects/3) tracking release. Once all those issues have been completed, we’ll have a 2.0 release date in sight!

**Comment by drwpow** [comment]
***Jan 2025 Update**

Terrazzo nearing release candidate status! It is now only blocked by some inflight changes to [the W3C Design Tokens format](https://www.designtokens.org/) which will block release. Here’s the release strategy:

1. DTCG finalizes new color format (support for CSS Color Module 4), Resolver spec (modes/themes), and a few other loose ends. ETA unknown, but possibly Mar 2025
1. Terrazzo will move from `0.x` → `2.0.0-rc.0` (release candidate), and continue version numbers from Cobalt
1. Cobalt will get a deprecation message to update to `@terrazzo/*` (the `@cobalt-ui/*` packages will still be available for use forever, but won’t receive many updates)

Below was the original message (with minor edits)

**Comment by drwpow** [comment]
Over 1 year later from this initial issue, we’re getting closer to a stable release! There’s a `2.0.0-alpha.0` release available that still has some bugs that we’re working through!

The 2.0.0 stable release will feature lots of additional features no one expected but I think will be welcome additions:

1. **Full support for `$ref`s** at any level—mix, match, reuse, and pull in remote JSON files in any way you like. This allows you to even do powerful things like duplicate token groups.
2. Full support for the **DTCG Resolver specification** (native modes & theming support)
    - Note: this will mean, sadly, `$extensions.mode` may go away in the future, in favor of the more “spec compliant“ approach. But `$extensions.mode` will work through all 2.0 versions at the very least, possibly longer.
3. **(Even more) advanced token linting**. In Cobalt, and earlier versions of Terrazzo, the ideas of validation and linting were separate. Now they are one!
    - Out-of-the-box, Terrazzo’s linter complies with the DTCG spec. But you have the option to opt out of specific validation checks, as well as add your own, to do what works best for you
4. **MCP support** (thanks to @Sidnioulz!). There’s some unknowns here, a lot of it is beyond me, but there will be _some_ story here for working with AI tools for DS management.

As of now there’s not an expected timeline for the 2.0.0 stable release. Just when I have time 🙂. But we’re nearing release candidate / final testing within the coming weeks, so, yay.

**Comment by drwpow** [comment]
A little over 2 years later, and 2.0.0 has released! I thiiiiiink we can probably unpin the issue now.

Go check out the docs on [terrazzo.app](https://terrazzo.app) which are now also rolled over to 2.0. Terrazzo now fully supports the stable version of DTCG, 2025.10, and can generate code for any project. Thanks for following along!
